### DISCUSSION TITLE
Inconsistent/incorrect GitHub Flavored Markdown linebreak rendering

### DISCUSSION THREAD

**Discussion by jrobsonchase** [community/community#64221] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

The GFM spec states:

> Paragraphs can contain multiple lines, but no blank lines:
> ```markdown
> aaa
> bbb
> 
> ccc
> ddd
> ```
> ```html
><p>aaa
> bbb</p>
> <p>ccc
> ddd</p>
> ```

Note the lack of hard line breaks within the paragraph tags. The [Hard line breaks](https://github.github.com/gfm/#hard-line-break) section lists some ways to get a hard line break to render within a paragraph, such as via a `\` or ending a line with two or more spaces.

So this
paragraph should
render
all in one
line. But instead, it
looks like
I'm writing some kind of visual
poetry.

Indeed, the HTML has line breaks rendered where there should be none:
```html
<p>So this<br>
paragraph should<br>
render<br>
all in one<br>
line. But instead, it<br>
looks like<br>
I'm writing some kind of visual<br>
poetry.</p>
```

However, this seems to work fine in markdown files checked into repositories. The source can be hard-wrapped to your heart's content, and the rendered version will omit the HTML line breaks.

It's especially frustrating when submitting issues or pull requests via some tool that allows editing in an environment that automatically does the hard-wrapping for prose, and links are involved. [This discussion](https://github.com/orgs/community/discussions/21065) touches on it, and the solution was to reflow commit messages for PR descriptions by removing the linebreaks in the source. [This older one](https://github.com/orgs/community/discussions/16848) comes to the same conclusion. Both say that GFM treats source line breaks as hard breaks in the output HTML, but the spec clearly makes the distinction between "soft" line breaks that _don't_ get rendered as HTML breaks, and "hard" line breaks that _do_. So the solution to reflow the markdown source to fit an out-of-spec renderer feels wrong.

**Comment by Jaber-Saed** [comment]
It seems like you're encountering an issue with GitHub Flavored Markdown (GFM) rendering line breaks in a way that's not consistent with the specification. The GFM spec does indeed distinguish between soft line breaks and hard line breaks, where soft line breaks are ignored and hard line breaks are converted to `<br>` tags. It's worth noting that different Markdown renderers might interpret line breaks slightly differently, leading to inconsistencies.

If you're encountering this issue when using GitHub's interface, such as when submitting issues or pull requests, here are a few things to consider:

1. **Markdown Rendering Engine**: Different platforms and tools might use slightly different Markdown rendering engines, which can lead to differences in how line breaks are handled. GitHub uses its own rendering engine for GFM.

2. **Editor Behavior**: Sometimes, the behavior of the Markdown editor itself can affect how line breaks are rendered. Some editors might interpret certain sequences of newlines as intentional line breaks, even if they're not strictly adhering to the Markdown specification.

3. **Workarounds**: While it's frustrating that the rendering behavior might not match the strict GFM specification, in cases where the rendered output is important (e.g., in issue descriptions, PR descriptions), you might need to adjust your source Markdown to work around the rendering quirks. This could involve removing or adjusting certain line breaks or using different formatting to achieve the desired layout.

4. **Feedback to GitHub**: If you believe that there's an inconsistency between the GFM spec and GitHub's rendering behavior, consider providing feedback to GitHub. They might be able to clarify their rendering logic or address the issue if it's indeed a bug.

5. **Markdown Linting**: If you're working on important documentation or content, consider using a Markdown linter or validation tool before submitting. This can help catch rendering discrepancies early.

Remember that Markdown rendering can sometimes be influenced by a combination of factors, including the rendering engine, the editor or interface, and the specific Markdown syntax used. In cases where the rendered output is critical, manual adjustments might be necessary to ensure consistent and desired results.

**Reply by jrobsonchase** [reply]
> 1. Markdown Rendering Engine: Different platforms and tools might use slightly different Markdown rendering engines, which can lead to differences in how line breaks are handled. GitHub uses its own rendering engine for GFM.

Yes, and GitHub's own rendering engine for GFM is out-of-spec for issues, pull requests, comments, and generally anything _normally_ submitted through the UI.

> 2. Editor Behavior: Sometimes, the behavior of the Markdown editor itself can affect how line breaks are rendered. Some editors might interpret certain sequences of newlines as intentional line breaks, even if they're not strictly adhering to the Markdown specification.

I submitted this discussion via the GitHub UI using their own editor. If you view the source of one of your own comments, you can see that the soft breaks aren't sneakily being converted to hard breaks or anything like that.

> 3. Feedback to GitHub: If you believe that there's an inconsistency between the GFM spec and GitHub's rendering behavior, consider providing feedback to GitHub. They might be able to clarify their rendering logic or address the issue if it's indeed a bug.

Yep, that's what this is and why I marked it as a Bug :+1: 

Another interesting thing I just noticed: If you "Quote Reply", it converts the comment back to markdown. But inexplicably, it removes the soft line breaks. My "visual poetry" paragraph quotes like this:

```markdown
> 
> So this paragraph should render all in one line. But instead, it looks like I'm writing some kind of visual poetry.
> 
```

I expected to get the exact markdown source from the quoted comment, but it appears to go through a prettifier that obeys the GFM spec better than their usual renderer does.

**Comment by Ryder17z** [comment]
i suppose this is why it's kind of impossible to get linebreaks in an issue comment?

**Comment by WRSomsky** [comment]
I'm seeing this as well.

I'm trying to find a way to preview (among other things) the `README.md` file for my repos.  I'm finding the `gh` extension `yusukebe/gh-markdown-preview` to be very useful, and it *almost* does everthing right, *except* that it presents newlines (with no trailing spaces, backslashes, *etc.*) as hard linebreaks, while when the `README.md` is uploaded to GitHub, those newlines are not interpreted that way and the text flows as a continuous paragraph.

I believe this behavior can be traced to the API, because when I use the command
```
gh api markdown -F mode=gfm -F text=@README.md
```
the results have embedded `<br>` HTML tags where the newlines were, which will cause the errant behavior.

Interestingly, if one uses the *non*-GFM mode of the API,
```
gh api markdown -F text=@README.md
```
then there are no embedded `<br>` tags. But, of course, the results then differ from the expected GFM-style interpretation in numerous other ways.

So there is ***definitely*** a discrepancy between the markdown translations of the API and the website.

**Comment by yugenelee** [comment]
There are two separate parts to this behaviour:

1. How the newline is parsed
2. How the resulting break is rendered

A normal newline without trailing spaces or a backslash is parsed as a **soft line break**:

```markdown
This is one
continuous paragraph.
```

It would normally produce HTML resembling:

```html
<p>This is one
continuous paragraph.</p>
```

Because the newline remains inside the same paragraph, the browser collapses it as ordinary whitespace and displays one continuous line of prose.

A **hard line break** is different. It is normally created using two trailing spaces or a backslash:

```markdown
This line ends with a hard break.\
This line begins underneath it.
```

That produces:

```html
<p>This line ends with a hard break.<br>
This line begins underneath it.</p>
```

The distinction between [soft and hard Markdown line breaks](https://markdific.com/docs/markdown/line-breaks/) is clear at the parsing level. However, the [GFM specification](https://github.github.com/gfm/#soft-line-breaks) explicitly permits a renderer to convert soft breaks into hard breaks.

GitHub’s Markdown implementation also exposes a `hardbreaks` rendering option. When enabled, soft-break nodes are emitted as `<br>` elements instead of ordinary newlines. This option can be seen in GitHub’s [Markdown rendering code](https://github.com/github/markup/blob/master/lib/github/markup/markdown.rb).

This appears to explain the difference between GitHub surfaces:

- Committed Markdown files such as `README.md` render soft breaks normally, so hard-wrapped source flows as one paragraph.
- Issues, pull requests, comments and discussions display soft breaks as visible line breaks.
- The Markdown API with `mode=gfm` appears to use the second behaviour, which is why it returns `<br>` elements.
- Calling the API without GFM mode avoids those inserted breaks, but it also changes other GFM-specific rendering behaviour.

This is therefore better described as a **rendering inconsistency between GitHub contexts**, rather than a Markdown parsing error. The newline can still be parsed as a soft break while being rendered as `<br>` by a particular configuration.

It also means that this command:

```bash
gh api markdown -F mode=gfm -F text=@README.md
```

cannot reliably preview how a committed `README.md` will look. It reproduces the API’s GFM configuration, which currently behaves more like GitHub’s conversational content.

Ideally, the API would provide either:

- A context that explicitly matches repository Markdown files
- An option to control whether soft breaks are rendered as `<br>`

The possible workarounds all have limitations:

- Use the API without `mode=gfm`, but lose some GFM-specific behaviour.
- Use a local CommonMark or GFM renderer with the necessary extensions enabled and `hardbreaks` disabled.
- Reflow hard-wrapped prose before sending it to the API, although this can alter lists, blockquotes and other Markdown structures.
- Avoid removing every `<br>` from the returned HTML, because that would also remove intentional hard breaks.

In short, the original observation is correct: GitHub renders identical soft breaks differently in repository files and conversational content. The GFM specification allows this renderer option, but the lack of parity, or an API setting that matches repository rendering, makes accurate README previews unnecessarily difficult.
